### PR TITLE
Jetpack Connect: query single site pre-connection

### PR DESCRIPTION
**Pull request by roccotripaldi**
Currently, when connecting a Jetpack site we wait for the query to `me/sites` to finish before allowing folks to start the process. We wait to ensure that the site we are connecting is not already on the site's list, or in the case of re-connection, that the site is removed before proceeding.

We can speed up the wait time by only querying the site in question.

To test reconnections:

- Disconnect a connected Jetpack site via the site's wp-admin
- Ensure that you can reconnect

Try the Jetpack Connect flows, and ensure there are no regressions. Ensure there is no long wait time when you land on the authorization screen.

### CODE REVIEW

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 1]
```diff
@@ -478,10 +478,10 @@ const LoggedInForm = React.createClass( {
 	renderStateAction() {
 		const { authorizeSuccess, siteReceived } = this.props.jetpackConnectAuthorize;
 		if (
-			this.props.isFetchingSites ||
+			this.props.isFetchingSite ||
 			this.isAuthorizing() ||
 			this.retryingAuth ||
-			( authorizeSuccess && ! siteReceived )
+			( this.props.isFetchingSites && authorizeSuccess && ! siteReceived )
```
Comment (sirreal) [inline client/signup/jetpack-connect/authorize-form.jsx]: These conditions seem confusing to me. After thinking about it for a long time, I think what they mean is:

* `this.props.isFetchingSite`: We're trying to fetch the requested auth site
* `( this.props.isFetchingSites && authorizeSuccess && ! siteReceived )`: The JP connect request has been authorized, but we haven't received the site data yet (from where?) and we're still waiting on global sites state to be updated.

If that third condition is true, we'll wind up with one of these messages, is that correct and the desired behavior?

```js
if ( authorizeSuccess && isRedirectingToWpAdmin ) {
	return this.translate( 'Returning to your site' );
}

if ( authorizeSuccess ) {
	return this.translate( 'Finishing up!', {
		context: 'Shown during a jetpack authorization process, while we retrieve the info we need to show the last page'
	} );
}
```

Something just seems off to me about making this third condition more strict. Maybe if you explain in English the motivation I'll feel better about it.

Comment (sirreal) [inline client/signup/jetpack-connect/authorize-form.jsx]: It feels a little off to make this condition more strict, can you explain the reasoning?

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 2]
```diff
@@ -628,6 +628,7 @@ export default connect(
 			jetpackSSOSessions: getSSOSessions( state ),
 			isAlreadyOnSitesList: isRemoteSiteOnSitesList( state ),
 			isFetchingSites: isRequestingSites( state ),
+			isFetchingSite: isRequestingSite( state ),
```
Comment (roccotripaldi) [inline client/signup/jetpack-connect/authorize-form.jsx]: `isFetchingAuthorizationSite` is good, i'll switch it for that.
With #13758, we were able to drop `isFetchingSites` entirely, but I'm trying to take smaller steps to solve the speed issue, so unfortunately, we do need both selectors.  But perhaps in a future PR, we can drop the second `me/sites` query as well, and make it a little clearer.

File: client/signup/jetpack-connect/authorize-form.jsx (hunk 1 above)
Comment (roccotripaldi) [inline client/signup/jetpack-connect/authorize-form.jsx]: You're right, it does not need to be so strict. I will fix.

**Comment by roccotripaldi** [comment]
> I ran into an issue. When connecting a new site and not logged in (2fa must be disabled), I consistently get the warning "Error authorizing. Page is refreshing for an other attempt." on my first attempt, and must approve a second time "In some cases, authorization can take a few attempts. Please try again."

@sirreal - I think this is related to the development-only login for on calypso.localhost:3000 - I've noticed that it happens when a test many PR's locally ( as specially poopy.life sites ). I will try to figure out why this is happening but I'm 95% sure that this PR is not introducing the error.

**Comment by roccotripaldi** [comment]
>  I'm 95% sure that this PR is not introducing the error.

I'm now 100% sure. I'm getting the same error on master, and I've reported here: #14145

**Comment by roccotripaldi** [comment]
I've rebased and addressed feedback. Test away please and thank you.

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 3]
```diff
@@ -150,6 +150,7 @@ export default connect(
 			jetpackSSOSessions: getSSOSessions( state ),
 			isAlreadyOnSitesList: isRemoteSiteOnSitesList( state ),
 			isFetchingSites: isRequestingSites( state ),
+			isFetchingAuthorizationSite: isRequestingSite( state ),
```
Comment (oskosk) [inline client/signup/jetpack-connect/authorize-form.jsx]: `isRequestingSite()` needs a siteId here.

Comment (roccotripaldi) [inline client/signup/jetpack-connect/authorize-form.jsx]: good catch! Thanks.

**Comment by oskosk** [comment]
LGTM!

File: client/state/jetpack-connect/reducer.js [hunk 4]
```diff
@@ -221,6 +222,12 @@ export function jetpackConnectAuthorize( state = {}, action ) {
 					bearerToken: action.data.bearer_token
 				}
 			);
+		case SITE_REQUEST_FAILURE:
+			const { client_id } = state.queryObject;
```
Comment (gwwar) [inline client/state/jetpack-connect/reducer.js]: @roccotripaldi do you mind following up with another PR to check if `queryObject` is available. The default state is `{}` so we may see the following. I notice this when I visit the purchases page, but I also have a lot of messed up test sites :)

screen shot 2017-05-22 at 5 36 19 pm

Comment (roccotripaldi) [inline client/state/jetpack-connect/reducer.js]: Thanks! PR opened: #14359
